### PR TITLE
Allow imports of JSON

### LINKED ISSUE #1048: Ability to import JSON
TypeScript with the `--resolveJsonModule` compiler option allows constructs like:

```ts
import * as foo from "./foo.json";
```

Where the imported `foo` namespace will match the structure of the `./foo.json` and ensure that the JSON is consumed in a type safe way.  TypeScript simply re-emit the JSON though and not actually make it importable.  In order to make usable within a Deno program we would have to create a AMD module something like this I would expect:

```js
define([], function() {
  return JSON.parse(/* fetched source code */);
});
```

### PR DESCRIPTION
**Pull request by kitsonk**
Fixes #1048 

This is a WIP PR for allowing the import of JSON as a module, which would include the ability to ensure its type safety in consuming code.  This would allow:

**config.json**

```json
{
  "debug": true,
  "version": "1.0.0"
}
```

**main.ts**

```ts
import * as configJson from "./config.json";

if (configJson.debug === true) {
  console.log(`Version: ${configJson.version}`);
}
```

### CODE REVIEW

**Comment by kitsonk** [comment]
@ry this is ready for review

File: js/compiler.ts [hunk 1]
```diff
@@ -196,8 +197,16 @@ export class DenoCompiler
         "Cannot run module without factory."
       );
       assert(moduleMetaData.hasRun === false, "Module has already been run.");
+      // while TypeScript always transpiles to modify the export object, to work
+      // with all types of AMD modules, we have to detect a return value from
+      // the factory
```
Comment (ry) [inline js/compiler.ts]: Move this comment --v

File: js/compiler.ts [hunk 2]
```diff
@@ -196,8 +197,16 @@ export class DenoCompiler
...
       // asserts not tracked by TypeScripts, so using not null operator
-      moduleMetaData.factory!(...this._getFactoryArguments(moduleMetaData));
+      const exports = moduleMetaData.factory!(
+        ...this._getFactoryArguments(moduleMetaData)
+      );
+      if (exports != null) {
```
Comment (ry) [inline js/compiler.ts]: Move above comment here. Mention that this is for JSON support (and potentially other things in the future)/

File: js/compiler.ts [hunk 3]
```diff
@@ -407,43 +416,60 @@ export class DenoCompiler
...
-      );
-      console.log(errMsg);
-      // All TypeScript errors are terminal for deno
-      this._os.exit(1);
-    }
+    // Instead of using TypeScript to transpile JSON modules, we will just do
+    // it directly.
+    if (mediaType === MediaType.Json) {
+      // tslint:disable-next-line:max-line-length
+      moduleMetaData.outputCode = `define([], function() { return JSON.parse(\`${sourceCode}\`); });\n//# sourceURL=${fileName}`;
```
Comment (ry) [inline js/compiler.ts]: Add a separate function for this.
```
export function jsonAmdTemplate(jsonStr: string): string
```
Add a test in compiler_test that does something like this
```
import { jsonAmdTemplate } from "./compiler";

test(function testJsonAmdTemplate() {
  const code = jsonAmdTemplate('{ "hello": "world", "foo", "bar" }');
  const result = eval(code);
  /// test result.
});

```

File: js/compiler.ts [hunk 4]
```diff
@@ -407,43 +416,60 @@ export class DenoCompiler
...
+    } else {
+      // TypeScript is overly opinionated that only CommonJS modules kinds can
```
Comment (ry) [inline js/compiler.ts]: Add
```
assert(MediaType.Ts == mediaType)
```

File: js/compiler_test.ts [hunk 5]
```diff
@@ -110,6 +110,12 @@ const fooBazTsOutput = `define(["require", "exports", "./bar.ts"], function (req
 });
 //# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiYmF6LmpzIiwic291cmNlUm9vdCI6IiIsInNvdXJjZXMiOlsiZmlsZTovLy9yb290L3Byb2plY3QvZm9vL2Jhei50cyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiOzs7SUFDQSxPQUFPLENBQUMsR0FBRyxDQUFDLFlBQUcsQ0FBQyxDQUFDIiwic291cmNlc0NvbnRlbnQiOlsiaW1wb3J0IHsgZm9vIH0gZnJvbSBcIi4vYmFyLnRzXCI7XG5jb25zb2xlLmxvZyhmb28pO1xuIl19
 //# sourceURL=/root/project/foo/baz.ts`;
+
+const loadConfigSource = `import * as config from "./config.json";
+console.log(config.foo.baz);
+`;
+const configJsonSource = `{"foo":{"bar": true,"baz": ["qat", 1]}}`;
+const configJsonOutput = `define([], function() { return JSON.parse(\`${configJsonSource}\`); });\n//# sourceURL=/root/project/config.json`;
```
Comment (ry) [inline js/compiler_test.ts]: Use jsonAmdTemplate(configJsonSource)

File: js/compiler_test.ts [hunk 6]
```diff
@@ -374,6 +397,23 @@ function teardown() {
...
+    factory = f;
+  }
+
+  const code = jsonAmdTemplate(`{ "hello": "world", "foo": "bar" }`);
+  const result = eval(code);
+  assert(result == null);
+  assertEqual(deps && deps.length, 0);
+  assert(factory != null);
+  const factoryResult = factory();
+  assertEqual(factoryResult, { hello: "world", foo: "bar" });
```
Comment (ry) [inline js/compiler_test.ts]: Nice. Thanks.

**Comment by zekth** [comment]
```
import { debug } from "./config.json";

if (debug === true) {
  console.log(`Version: ${configJson.version}`);
}
```

This output :
```
The requested module './config.json' does not provide an export named 'version'
```

Would it be possible to support it in the future?

**Pull request merged**